fvm/programs/programs.go

```diff
@@ -93,38 +93,23 @@ func (p *Programs) HasChanges() bool {
 	return len(p.programs) > 0 || p.cleaned
 }
 
-func (p *Programs) unsafeForceCleanup() {
-	p.cleaned = true
-
-	// Stop using parent's data to prevent
-	// infinite chaining of objects
-	p.parent = nil
-
-	// start with empty storage
-	p.programs = make(map[common.LocationID]*ProgramEntry)
-}
-
-// ForceCleanup is used to force a complete cleanup
-// It exists temporarily to facilitate a temporary measure which can retry
-// a transaction in case checking fails
-// It should be gone when the extra retry is gone
-func (p *Programs) ForceCleanup() {
-	p.lock.Lock()
-	defer p.lock.Unlock()
-
-	p.unsafeForceCleanup()
-}
-
-func (p *Programs) Cleanup(changedContracts []ContractUpdateKey) {
+func (p *Programs) Cleanup(changedContracts []ContractUpdateKey, hasFreezedAccounts bool) {
 	p.lock.Lock()
 	defer p.lock.Unlock()
 
 	// In mature system, we would track dependencies between contracts
 	// and invalidate only affected ones, possibly setting them to
 	// nil so they will override parent's data, but for now
 	// just throw everything away and use a special flag for this
-	if len(changedContracts) > 0 {
-		p.unsafeForceCleanup()
+	if len(changedContracts) > 0 || hasFreezedAccounts {
```

Comment:
Here we use a list of contracts despite the fact that we only check the length. This is in an anticipation that we will use this information one day to properly invalidate affected entire only.
We should probably do the same for accounts, ie provide it with a list of frozen accounts rather than set a boolean flag.